File: trip_capstone_final/src/main/java/com/example/service/TravelPackageService.java
package com.example.service;

import com.example.dto.TravelPackageDTO;
import com.example.entity.TravelPackageEntity;
import com.example.exceptions.IdNotFoundException;
import com.example.mapper.TravelPackageMapper;
import com.example.repository.TravelPackageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.sql.Date;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TravelPackageService {

    @Autowired
    private TravelPackageRepository travelPackageRepository;

    @Autowired
    private TravelPackageMapper travelPackageMapper;

    public List<TravelPackageDTO> getAllPackages() {
        return travelPackageRepository.findAll().stream()
                .map(travelPackageMapper::toDTO)
                .collect(Collectors.toList());
    }

    public List<TravelPackageDTO> getAllActivePackages() {
        return travelPackageRepository.findByPackageStatus("ACTIVE").stream()
                .map(travelPackageMapper::toDTO)
                .collect(Collectors.toList());
    }

    public TravelPackageDTO getPackageById(Long id) {
        TravelPackageEntity entity = travelPackageRepository.findById(id)
                .orElseThrow(() -> new IdNotFoundException("TravelPackage", "id", id));
        return travelPackageMapper.toDTO(entity);
    }

    public List<TravelPackageDTO> searchPackages(String source, String destination, Date fromDate, Date toDate) {
        return travelPackageRepository.searchPackages(source, destination, fromDate, toDate).stream()
                .map(travelPackageMapper::toDTO)
                .collect(Collectors.toList());
    }

    public TravelPackageDTO addPackage(TravelPackageDTO dto) {
        TravelPackageEntity saved = travelPackageRepository.saveAndFlush(travelPackageMapper.toEntity(dto));
        return travelPackageMapper.toDTO(saved);
    }

    public TravelPackageDTO updatePackage(TravelPackageDTO dto, Long id) {
        TravelPackageEntity existing = travelPackageRepository.findById(id)
                .orElseThrow(() -> new IdNotFoundException("TravelPackage", "id", id));
        TravelPackageEntity updated = travelPackageMapper.toEntity(dto);
        updated.setId(existing.getId());
        return travelPackageMapper.toDTO(travelPackageRepository.save(updated));
    }

    public void deletePackage(Long id) {
        TravelPackageEntity entity = travelPackageRepository.findById(id)
                .orElseThrow(() -> new IdNotFoundException("TravelPackage", "id", id));
        travelPackageRepository.delete(entity);
    }
}

File: frontend/src/app/services/travel-package-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { TravelPackageDTO } from '../models/TravelPackageDTO.model';

@Injectable({ providedIn: 'root' })
export class TravelPackageService {
  private baseUrl = 'http://localhost:9090/api/travel-packages'; // matches Spring Boot

  constructor(private httpClient: HttpClient) {}

  getAllPackages(): Observable<TravelPackageDTO[]> {
    return this.httpClient.get<TravelPackageDTO[]>(this.baseUrl); // GET /api/travel-packages
  }

  getAllActivePackages(): Observable<TravelPackageDTO[]> {
    return this.httpClient.get<TravelPackageDTO[]>(`${this.baseUrl}/active`); // GET /api/travel-packages/active
  }

  getPackageById(id: number): Observable<TravelPackageDTO> {
    return this.httpClient.get<TravelPackageDTO>(`${this.baseUrl}/${id}`); // GET /api/travel-packages/{id}
  }

  searchPackages(
    source: string,
    destination: string,
    fromDate: string,
    toDate: string
  ): Observable<TravelPackageDTO[]> {
    return this.httpClient.get<TravelPackageDTO[]>(
      `${this.baseUrl}/search?source=${source}&destination=${destination}&fromDate=${fromDate}&toDate=${toDate}`
    ); // GET /api/travel-packages/search
  }

  addPackage(dto: TravelPackageDTO): Observable<TravelPackageDTO> {
    return this.httpClient.post<TravelPackageDTO>(this.baseUrl, dto); // POST /api/travel-packages
  }

  updatePackage(dto: TravelPackageDTO, id: number): Observable<TravelPackageDTO> {
    return this.httpClient.put<TravelPackageDTO>(`${this.baseUrl}/${id}`, dto); // PUT /api/travel-packages/{id}
  }

  deletePackage(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.baseUrl}/${id}`); // DELETE /api/travel-packages/{id}
  }
}
